src/harness/test_harness_objects.py

```diff
@@ -0,0 +1,324 @@
+    heartbeat_requests, heartbeat_responses = self.heartbeatRequestForAllActiveGrants()
+    relinquishment_requests = []
+    grant_requests = []
+
+    # Construct relinquishment and grant for the heart beat response which have operation param
+    for heartbeat_response, heartbeat_request in zip(heartbeat_responses, heartbeat_requests):
+        if heartbeat_response['response']['responseCode'] == Grant.ResponseCodes.TERMINATED_GRANT.value:
+           if 'operationParam' in heartbeat_response:
+             # Construct grant request for Terminated grants
+             grant_requests.append(self._constructGrantRequests(heartbeat_request))
+           else:
+             logging.error('Invalid Response since operation param not present in case of Terminated Grant')
+
+        elif heartbeat_response['response']['responseCode'] != 0:
+          # If any heartbeat contain suggested operation params,construct relinquishment requests
+          if 'operationParam' in heartbeat_response:
+            relinquishment_request = {
+                'cbsdId': heartbeat_response['cbsdId'],
+                'grantId': heartbeat_response['grantId']
+            }
+            relinquishment_requests.append(relinquishment_request)
+
+            # Construct grant request for relinquished grants
+            grant_requests.append(self._constructGrantRequests(heartbeat_request))
+
+            # Delete the grant object
+            del self.cbsd_objects[heartbeat_response['cbsdId']].grant_objects[heartbeat_response['grantId']]
+
+    # Perform relinquishment since operation param present in heartbeat response
+    if len(relinquishment_requests):
+      relinquishment_requests_wrap = {
+        'relinquishmentRequest': relinquishment_requests
+      }
+      relinquishment_responses = self.testcase._sas.Relinquishment(relinquishment_requests_wrap,
+                                        self.ssl_cert, self.ssl_key)['relinquishmentResponse']
+
+      # Validate the response of the relinquishment request
+      for relinquishment_request, relinquishment_response in \
+              zip(relinquishment_requests, relinquishment_responses):
+        self.testcase.assertEqual(relinquishment_response['response']['responseCode'], 0)
+        self.testcase.assertEqual(relinquishment_response['cbsdId'], relinquishment_request['cbsdId'])
+        self.testcase.assertEqual(relinquishment_response['grantId'], relinquishment_request['grantId'])
+
+    # Check if any grant request need to be sent
+    if len(grant_requests):
+      grant_requests_wrap = {
+        'grantRequest': grant_requests
+      }
+      # Perform Grant for relinquished grants
+      grant_responses = self.testcase._sas.Grant(grant_requests_wrap,
+                                                 self.ssl_cert, self.ssl_key)['grantResponse']
+
+      for grant_response, grant_request in zip(grant_responses, grant_requests):
+        cbsd_object = self.cbsd_objects[grant_request['cbsdId']]
```

Comment:
What happens if the Grant request fails?

Reply:
We ignore the grants failure per Test dev F2F working doc. Do we need adding handling logic ?

Reply:
Yes, because the next line will raise an exception (invalid key) if the grant request failed. So code *should* be ignoring the failure, but the current code is not actually ignoring it.

Reply:
Changed   as per suggestion.